### PR TITLE
DC-2842

### PR DESCRIPTION
**Pull request by jim-collins**

### CODE REVIEW

**Review by OlegEfrem — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/securemessage/connectors/ChannelPreferencesConnector.scala [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      case JsString(s) =>
+        Try(EmailAddress(s)) match {
+          case Success(v) => JsSuccess(v)
+          case Failure(e) => JsError(e.getMessage)
+        }
+      case _ => JsError("Uable to parse email address")
+    }
+  }
+
+  private def parseEmail(body: String): Either[Int, EmailAddress] =
```
Comment (OlegEfrem) [inline app/uk/gov/hmrc/securemessage/connectors/ChannelPreferencesConnector.scala:49]: You probably want to return as left an EmailLookupException(code: Int, message: String)

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala [hunk 2]
```diff
@@ -19,50 +19,70 @@ package uk.gov.hmrc.securemessage.services
...
-  private def getEmailRequest(request: ConversationRequest): Either[String, EmailRequest] = {
-    val emailAddresses = request.recipients.flatMap(r => r.customer.email.toList)
-    if (emailAddresses.nonEmpty) {
-      Right[String, EmailRequest](EmailRequest(emailAddresses, request.alert.templateId))
-    } else {
-      Left[String, EmailRequest]("No recipient email addresses provided")
-    }
-  }
+  private def lookupMissingCustomerEmails(cnv: Conversation)(implicit hc: HeaderCarrier,
+                                                             ec: ExecutionContext): Future[(MissingEmailAddress, Conversation)] =
```
Comment (OlegEfrem) [inline app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:70]: you probably just propagate as left here the EmailLookupException

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala [hunk 3]
```diff
@@ -19,50 +19,70 @@ package uk.gov.hmrc.securemessage.services
...
+  private val EmailNotFound: MissingEmailAddress = !EmailFound
+  private val emailNotFoundErrorMsg = Json.parse("""{"error": 424, "message": "Verified email address could not be found"}""")
 
   def createConversation(conversationRequest: ConversationRequest, client: String, conversationId: String)(
     implicit hc: HeaderCarrier,
-    ec: ExecutionContext): Future[Result] = {
-    val conversation = conversationRequest.asConversation(client, conversationId)
-    repo.insertIfUnique(conversation).map { isUnique =>
+    ec: ExecutionContext): Future[Result] =
+    for {
```
Comment (OlegEfrem) [inline app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:51]: you wanna do for comprehension on Either and return an Either as well, propagating the error.

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala [hunk 4]
```diff
@@ -19,50 +19,70 @@ package uk.gov.hmrc.securemessage.services
...
       if (isUnique) {
-        getEmailRequest(conversationRequest) match {
-          case Left(err) => BadRequest(err)
-          case Right(req) =>
-            val _ = emailConnector.send(req)
-            Created
-        }
+        val _ = emailConnector.send(EmailRequest(
+          cnv.participants.filter(p => p.participantType == PCustomer).map(_.email.get), cnv.alert.templateId))
+        Created
```
Comment (OlegEfrem) [inline app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:63]: no play statuses should be here.

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala [hunk 5]
```diff
@@ -19,50 +19,70 @@ package uk.gov.hmrc.securemessage.services
...
       } else {
         Conflict("Duplicate of existing conversation")
```
Comment (OlegEfrem) [inline app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:65]: you probably want to change this to Left(ConversationException(code: Int, message: String)

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala (hunk 3 above)
Comment (OlegEfrem) [reply app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:51]: another question here is: if one email lookup fails, do we fail all ?

File: app/uk/gov/hmrc/securemessage/connectors/ChannelPreferencesConnector.scala [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.{ Failure, Success, Try }
+
+@Singleton
+@SuppressWarnings(Array("org.wartremover.warts.All"))
+class ChannelPreferencesConnector @Inject()(config: Configuration, httpClient: HttpClient)(
+  implicit ec: ExecutionContext)
+    extends ServicesConfig(config) with Logging {
+
+  implicit object EmailAddressReads extends Reads[EmailAddress] {
```
[the remark this answers is not in the document]
Comment (jim-collins) [reply app/uk/gov/hmrc/securemessage/connectors/ChannelPreferencesConnector.scala:38]: I did not see that I will use is.

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala [hunk 7]
```diff
@@ -19,50 +19,70 @@ package uk.gov.hmrc.securemessage.services
...
+        channelPrefConnector.getEmailForEnrolment(b.identifier).map({
+        case Right(ea) => (EmailFound, b.copy(email = Some(ea)))
+        case _ => (EmailNotFound, b)
+      }
+      ) :: z }
+    )).map(l =>
+      l.foldLeft[(MissingEmailAddress, List[Participant])]((EmailFound, List.empty[Participant]))(
+        (z, b) => (z._1 && b._1, b._2 :: z._2)
+      )
+    ).map(f => (f._1, cnv.copy(participants = f._2)))
```
Comment (OlegEfrem) [inline app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:85]: the whole function is quite complicated, can we move a single email lookup into another function, then just map on the participants list ?

File: app/uk/gov/hmrc/securemessage/connectors/ChannelPreferencesConnector.scala [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+import uk.gov.hmrc.http.HttpReads.Implicits._
+import uk.gov.hmrc.play.bootstrap.config.ServicesConfig
+import uk.gov.hmrc.securemessage.models.core.Identifier
+
+import javax.inject.{ Inject, Singleton }
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.{ Failure, Success, Try }
+
+@Singleton
+@SuppressWarnings(Array("org.wartremover.warts.All"))
```
Comment (elabeca) [inline app/uk/gov/hmrc/securemessage/connectors/ChannelPreferencesConnector.scala:33]: just wondering, do we really need to suppress all wartremover warnings or was that just left over?

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala [hunk 9]
```diff
@@ -19,50 +19,70 @@ package uk.gov.hmrc.securemessage.services
...
 
-import scala.concurrent.{ ExecutionContext, Future }
+import scala.concurrent.{ExecutionContext, Future}
 /*
 TODO: refactor so that service has no play dependencies, only core classes.
  */
 
-@SuppressWarnings(Array("org.wartremover.warts.ImplicitParameter"))
-class SecureMessageService @Inject()(repo: ConversationRepository, emailConnector: EmailConnector) {
+@SuppressWarnings(Array("org.wartremover.warts.All"))
```
Comment (elabeca) [inline app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:39]: same here, removing all wartremover warnings?

File: test/uk/gov/hmrc/securemessage/connectors/ChannelPreferencesConnectorSpec.scala [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+import play.api.http.Status.{ BAD_GATEWAY, BAD_REQUEST, OK }
+import uk.gov.hmrc.emailaddress.EmailAddress
+import uk.gov.hmrc.securemessage.models.core.Identifier
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration.Duration
+import scala.concurrent.{ Await, Future }
+
+@SuppressWarnings(Array("org.wartremover.warts.All"))
```
Comment (elabeca) [inline test/uk/gov/hmrc/securemessage/connectors/ChannelPreferencesConnectorSpec.scala:35]: same here regarding wartremover

File: test/uk/gov/hmrc/securemessage/connectors/ChannelPreferencesConnectorSpec.scala [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+  private val validEmailVerification = """{"address":"some@email.com","timestamp":"1987-03-20T01:02:03.000Z"}"""
+  private val inValidEmailVerification = """{"add":"some@email.com","timestamp":"1987-03-20T01:02:03.000Z"}"""
+  private val inValidJson = """{"add:"some@email.com","timestamp":"1987-03-20T01:02:03.000Z"}"""
+  val cpURL =
+    "https://host:443/channel-preferences/preference/email?enrolmentKey=HMRC-CUS-ORG&taxIdName=EORINumber&taxIdValue=eoriid"
+
+  "getEmailForEnrolment" should {
+    "return the email address if found by CDS" in new TestCase {
+      private val connector = new ChannelPreferencesConnector(configuration, mockHttpClient)
+      when(
```
Comment (OlegEfrem) [inline test/uk/gov/hmrc/securemessage/connectors/ChannelPreferencesConnectorSpec.scala:48]: shall we move mocks and test data to specialised TestContext/TestCase?
see example: https://github.com/hmrc/secure-message/blob/master/test/uk/gov/hmrc/securemessage/controllers/SecureMessageControllerSpec.scala#L56

File: test/uk/gov/hmrc/securemessage/services/SecureMessageServiceSpec.scala [hunk 12]
```diff
@@ -55,10 +56,19 @@ class SecureMessageServiceSpec extends PlaySpec with ScalaFutures with MockitoSu
       private val result = service.createConversation(ConversationUtil.getConversationRequest(true), "cdcm", "123")
       status(result) mustBe CREATED
     }
-    "return BAD REQUEST (400) when no email address is provided" in new TestCase {
+    "return FAILED DEPENDENCY (424) when no email address is provided" in new TestCase {
```
Comment (OlegEfrem) [inline test/uk/gov/hmrc/securemessage/services/SecureMessageServiceSpec.scala:59]: shall we move mocks and test data to specialised TestContext/TestCase?
see example: https://github.com/hmrc/secure-message/blob/master/test/uk/gov/hmrc/securemessage/controllers/SecureMessageControllerSpec.scala#L56

**Review by OlegEfrem — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala [hunk 13]
```diff
@@ -54,13 +60,15 @@ class ConversationRepository @Inject()(implicit connector: MongoConnector)
...
     insert(conversation)
-      .map(_ => true)
+      .map(_ => Right(true))
       .recoverWith {
         case e: DatabaseException if e.code.contains(DuplicateKey) =>
-          logger.warn("Ignoring duplicate found on insertion to conversation collection: " + e.getMessage())
-          Future.successful(false)
+          val errMsg = "Ignoring duplicate found on insertion to conversation collection: " + e.getMessage()
+          logger.error(errMsg, e)
+          Future.successful(Left(StoreException(CONFLICT, errMsg, Some(e))))
```
Comment (OlegEfrem) [inline app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala:71]: can we have a specialised exeption: DuplicateConversationException(message, error) extends StoreException. 
trait StoreException {
    def message: String
    def exception: Throwable
}

**Review by OlegEfrem — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/securemessage/controllers/SecureMessageController.scala [hunk 14]
```diff
@@ -38,7 +42,12 @@ class SecureMessageController @Inject()(
   def createConversation(client: String, conversationId: String): Action[JsValue] = Action.async(parse.json) {
     implicit request =>
       withJsonBody[ConversationRequest] { conversationRequest =>
-        secureMessageService.createConversation(conversationRequest, client, conversationId)
+        secureMessageService.createConversation(conversationRequest.asConversation(client, conversationId)).map {
+          case Right(_)                          => Created
+          case Left(ee: EmailException)          => Created(ee.message)
+          case Left(se: StoreException)          => Conflict(se.message)
```
Comment (OlegEfrem) [inline app/uk/gov/hmrc/securemessage/controllers/SecureMessageController.scala:48]: Here we need to match on specialised DuplicateConversationException, as there might be other store exceptions which should not be mapped to Conflict http response code.

File: app/uk/gov/hmrc/securemessage/controllers/SecureMessageController.scala [hunk 15]
```diff
@@ -38,7 +42,12 @@ class SecureMessageController @Inject()(
...
+          case Left(sme: SecureMessageException) => Created(sme.message)
```
Comment (OlegEfrem) [inline app/uk/gov/hmrc/securemessage/controllers/SecureMessageController.scala:49]: what are other exception ? 
I guess here should be case Left(error: Throwable) =>
   logger.error(error.message, error) 
   INTERNAL_SEVER_ERROR(error.message)

File: app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala (hunk 13 above)
Comment (jim-collins) [reply app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala:71]: StoreException was the one you created for this scenario

Comment (jim-collins) [reply app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala:71]: StoreException is a case class not a trait

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala [hunk 16]
```diff
@@ -19,51 +19,83 @@ package uk.gov.hmrc.securemessage.services
...
-    val emailAddresses = request.recipients.flatMap(r => r.customer.email.toList)
-    if (emailAddresses.nonEmpty) {
-      Right[String, EmailRequest](EmailRequest(emailAddresses, request.alert.templateId))
-    } else {
-      Left[String, EmailRequest]("No recipient email addresses provided")
-    }
-  }
+  @SuppressWarnings(Array("org.wartremover.warts.Equals", "org.wartremover.warts.ImplicitParameter"))
+  private def getConversationWithEmailFromCds(
+    cnv: Conversation)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Conversation] =
```
Comment (OlegEfrem) [inline app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:83]: shall this return a Future[Either[Exception, _] ?

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala [hunk 17]
```diff
@@ -19,51 +19,83 @@ package uk.gov.hmrc.securemessage.services
...
+          channelPrefConnector
+            .getEmailForEnrolment(p.identifier)
+            .map(
+              _.fold(_ => p, e => p.copy(email = Some(e)))
+            )
+        }
+      }))
+      .map(ps => cnv.copy(participants = ps))
 
+  @SuppressWarnings(Array("org.wartremover.warts.ImplicitParameter"))
```
Comment (OlegEfrem) [inline app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:98]: i guess if this is on the class level, doesn't need to be on function level as well, isn't it ?

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala [hunk 18]
```diff
@@ -19,51 +19,83 @@ package uk.gov.hmrc.securemessage.services
...
+  val hasEmail: Participant => Boolean = p =>
+    p.participantType == PSystem || (p.participantType == PCustomer && p.email.isDefined)
+
+  @SuppressWarnings(
+    Array(
+      "org.wartremover.warts.Nothing",
+      "org.wartremover.warts.Any",
+      "org.wartremover.warts.Equals",
+      "org.wartremover.warts.ImplicitParameter"))
+  private def emailParticipants(cnv: Conversation)(
```
Comment (OlegEfrem) [inline app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:61]: maybe we can change this function to sendEmail(customers: Seq[Participant]

**Review by OlegEfrem — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/securemessage/controllers/SecureMessageController.scala [hunk 19]
```diff
@@ -38,7 +42,12 @@ class SecureMessageController @Inject()(
   def createConversation(client: String, conversationId: String): Action[JsValue] = Action.async(parse.json) {
     implicit request =>
       withJsonBody[ConversationRequest] { conversationRequest =>
-        secureMessageService.createConversation(conversationRequest, client, conversationId)
+        secureMessageService.createConversation(conversationRequest.asConversation(client, conversationId)).map {
```
Comment (OlegEfrem) [inline app/uk/gov/hmrc/securemessage/controllers/SecureMessageController.scala:45]: love this clear pattern: 
call service, map result to specific Http Responses. 
Good job :)

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala (hunk 18 above)
Comment (OlegEfrem) [inline app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:61]: so many exceptions, can we change the code so there are no warts ?

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala [hunk 20]
```diff
@@ -19,51 +19,83 @@ package uk.gov.hmrc.securemessage.services
...
+    cnv: Conversation)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Conversation] =
+    Future
+      .sequence(cnv.participants.map(p => {
+        if (p.participantType == PSystem || (p.participantType == PCustomer && p.email.isDefined)) {
+          Future.successful(p)
+        } else {
+          channelPrefConnector
+            .getEmailForEnrolment(p.identifier)
+            .map(
+              _.fold(_ => p, e => p.copy(email = Some(e)))
```
Comment (OlegEfrem) [inline app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:92]: so here you will lose the unsuccessful lockups ?

File: app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala [hunk 21]
```diff
@@ -31,12 +32,17 @@ import uk.gov.hmrc.mongo.{ MongoConnector, ReactiveRepository }
...
 import uk.gov.hmrc.securemessage.models.core.Message.dateFormat
 import uk.gov.hmrc.securemessage.models.core.{ Conversation, Message, Participants }
+import uk.gov.hmrc.securemessage.services.exception.{ HttpException }
+
 import javax.inject.{ Inject, Singleton }
 import scala.collection.Seq
 import scala.concurrent.{ ExecutionContext, Future }
 
+final case class StoreException(code: Int, message: String, override val throwable: Option[Throwable])
+    extends HttpException
```
[the remark this answers is not in the document]
Comment (jim-collins) [reply app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala:42]: You did that!

File: app/uk/gov/hmrc/securemessage/services/exception/SecureMessageException.scala [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.securemessage.services.exception
+
+trait SecureMessageException {
```
Comment (OlegEfrem) [inline app/uk/gov/hmrc/securemessage/services/exception/SecureMessageException.scala:19]: i realised now, we need secure message exception to be a class and extend from java exception, so that in the controller we can handle the InternalSeevrError Htpp 500 with the last case _: Left(error: Exception) =>

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala (hunk 18 above)
Comment (jim-collins) [reply app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:61]: We pass implicts everywhere what is wrong with ==

File: test/uk/gov/hmrc/securemessage/controllers/SecureMessageControllerSpec.scala [hunk 23]
```diff
@@ -76,6 +78,35 @@ class SecureMessageControllerSpec extends PlaySpec with ScalaFutures with Mockit
...
+        .thenReturn(Future(Left(StoreException(CONFLICT, "conflict error", None))))
+      private val response = controller.createConversation("cdcm", "123")(fakeRequest)
+      status(response) mustBe CONFLICT
+      contentAsString(response) mustBe "conflict error"
+    }
+
+    "return CREATED (201) when there is an error sending the email" in new CreateConversationTestCase(
+      requestBody = Resources.readJson("model/api/create-conversation-full.json")) {
+      when(mockSecureMessageService.createConversation(any[Conversation])(any[HeaderCarrier], any[ExecutionContext]))
+        .thenReturn(Future(Left(EmailException(CREATED, "email error"))))
```
Comment (OlegEfrem) [inline test/uk/gov/hmrc/securemessage/controllers/SecureMessageControllerSpec.scala:94]: this makes me think we should remove code from email exception and assign an HTTP code at the controller level when we handle exceptions.

**Review by OlegEfrem — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala (hunk 20 above)
Comment (jim-collins) [reply app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:92]: Here I return the original participant with no email

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala (hunk 18 above)
Comment (jim-collins) [reply app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:61]: I have have changed it. There is still the same number of warts there is nothing wrong with the code.

File: test/uk/gov/hmrc/securemessage/connectors/ChannelPreferencesConnectorSpec.scala (hunk 11 above)
Comment (jim-collins) [reply test/uk/gov/hmrc/securemessage/connectors/ChannelPreferencesConnectorSpec.scala:47]: Done

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala [hunk 24]
```diff
@@ -19,51 +19,83 @@ package uk.gov.hmrc.securemessage.services
...
+    ec: ExecutionContext): Future[Either[SecureMessageException, Int]] =
+    (for {
+      cnv <- EitherT.right[SecureMessageException](getConversationWithEmailFromCds(conversation))
+      _   <- EitherT(repo.insertIfUnique(cnv))
+      res <- emailParticipants(cnv)
+    } yield res).value
 
-  def createConversation(conversationRequest: ConversationRequest, client: String, conversationId: String)(
+  @SuppressWarnings(Array("org.wartremover.warts.Equals"))
+  val hasEmail: Participant => Boolean = p =>
```
[the remark this answers is not in the document]
Comment (jim-collins) [reply app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:52]: Thanks Elie

File: app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala (hunk 13 above)
Comment (alexflav23) [reply app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala:71]: do we pass a throwable or root exception here too or just a msg?

File: app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala (hunk 21 above)
[the remark this answers is not in the document]
Comment (OlegEfrem) [reply app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala:42]: yeah, and i did it badly :D

File: app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala (hunk 13 above)
Comment (OlegEfrem) [reply app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala:71]: yes, and now i realise we need a better/more specialised exception :D

Comment (OlegEfrem) [reply app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala:71]: > do we pass a throwable or root exception here too or just a msg?

yes we should pass the parent exception, the type should be whatever is in java.lang.Exception, as that should be our base.

Comment (OlegEfrem) [reply app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala:71]: > StoreException was the one you created for this scenario

yes, and now i realise we need a more specialised exception :)

Comment (OlegEfrem) [reply app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala:71]: > StoreException is a case class not a trait

yes, and i suggest we have the types and  Hierarchy like like this: 
class SecureMessageException(message: String, cause: Throwable) extends Exception(message, cause) this should be the top level exception
class StoreException(message: String, cause: Throwable) extends SecureMessageException(message, cause) - top level for all DB exceptions
case class DuplicateConversationException(message: String, cause: Throwable) extends StoreException(message, cause) - specialised exception for duplicate conversation at db level.
Same pattern followed for other parts, ex: 
class ConnectorException(message: String, cause: Throwable) extends SecureMessageException(message, cause)
case class EmailLookupException(message: String, cause: Throwable) extends ConnectorException(message, cause)

Bottom line we should not have code in any of the exceptions, we should have specialised exeptions and map them to the HTTP codes in the controllers only. If an upsteam system responds with a non success http code, then our code will be 502 and the message will say: couldn't retrieve: what_data from upstream system x, return error code: xxx, with the error message: xxx.

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala (hunk 17 above)
Comment (jim-collins) [reply app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:98]: I was told to remove it from the class level and put it on the function.

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala (hunk 16 above)
Comment (jim-collins) [reply app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:83]: No it returns the Conversation either with the emails retrieved from CDS or the Conversation with the original participants.

File: app/uk/gov/hmrc/securemessage/services/exception/SecureMessageException.scala (hunk 22 above)
Comment (jim-collins) [reply app/uk/gov/hmrc/securemessage/services/exception/SecureMessageException.scala:19]: I have changed it to this. I don't see why we need to extend Exception
class SecureMessageException(val message: String, val throwable: Option[Throwable] = None)

File: test/uk/gov/hmrc/securemessage/controllers/SecureMessageControllerSpec.scala (hunk 23 above)
Comment (jim-collins) [reply test/uk/gov/hmrc/securemessage/controllers/SecureMessageControllerSpec.scala:94]: I have removed the code from all classes

File: app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala (hunk 13 above)
Comment (jim-collins) [reply app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala:71]: So you are saying now we need to create an instance of Throwable in every class.

Comment (OlegEfrem) [reply app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala:71]: you are right, removed extending from Exception, made cause option, and replace Exception suffix with Error.

File: test/uk/gov/hmrc/securemessage/controllers/SecureMessageControllerSpec.scala [hunk 25]
```diff
@@ -53,29 +55,84 @@ class SecureMessageControllerSpec extends PlaySpec with ScalaFutures with Mockit
 
   "createConversation" must {
 
-    "return CREATED (201) when sent a request with all optional fields populated" in new CreateConversationTestCase(
+    s"return CREATED ($CREATED) when sent a request with all optional fields populated" in new CreateConversationTestCase(
```
Comment (OlegEfrem) [inline test/uk/gov/hmrc/securemessage/controllers/SecureMessageControllerSpec.scala:58]: you can't right client and run a single test if you have string interpolations here, so better to revert this to what was before

File: test/uk/gov/hmrc/securemessage/services/SecureMessageServiceSpec.scala [hunk 26]
```diff
@@ -45,25 +50,47 @@ class SecureMessageServiceSpec extends PlaySpec with ScalaFutures with MockitoSu
...
   implicit val hc: HeaderCarrier = HeaderCarrier()
 
   "createConversation" must {
 
-    "return CREATED (201) when an email address is provided" in new TestCase {
-      when(mockRepository.insertIfUnique(any[Conversation])(any[ExecutionContext])).thenReturn(Future(true))
-      private val result = service.createConversation(ConversationUtil.getConversationRequest(true), "cdcm", "123")
-      status(result) mustBe CREATED
+    s"return CREATED ($CREATED) when an email address is provided" in new TestCase {
+      reset(mockRepository)
```
Comment (OlegEfrem) [inline test/uk/gov/hmrc/securemessage/services/SecureMessageServiceSpec.scala:61]: it is better to have mocks in test context, you don't need resets then. 
And good to be consistent, all mocks and test data should be in test context across all tests.

File: test/uk/gov/hmrc/securemessage/services/SecureMessageServiceSpec.scala [hunk 27]
```diff
@@ -45,25 +50,47 @@ class SecureMessageServiceSpec extends PlaySpec with ScalaFutures with MockitoSu
...
+        .thenReturn(Future(Left(EmailLookupError(""))))
+      private val result = service.createConversation(cnvWithNoEmail).futureValue
+      result.swap.toOption.get.message mustBe "Verified email address could not be found"
     }
-    "return CONFLICT (409) when a conversation already exists for this client and conversation ID" in new TestCase {
-      when(mockRepository.insertIfUnique(any[Conversation])(any[ExecutionContext])).thenReturn(Future(false))
-      private val result = service.createConversation(ConversationUtil.getConversationRequest(true), "cdcm", "123")
-      status(result) mustBe CONFLICT
+    s"return CREATED ($CREATED) when no email address is provided but is found in the CDS lookup" in new TestCase {
+      reset(mockRepository)
```
Comment (OlegEfrem) [inline test/uk/gov/hmrc/securemessage/services/SecureMessageServiceSpec.scala:76]: please move all mocks and test data to test context.

**Review by OlegEfrem — APPROVED** [review]

**Review by OlegEfrem — CHANGES_REQUESTED** [review]
request changes on the comments

**Review by elabeca — DISMISSED** [review]

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala [hunk 28]
```diff
@@ -107,6 +138,7 @@ class SecureMessageService @Inject()(repo: ConversationRepository, emailConnecto
     val participantIdentifiers = getIdentifiersFromParticipants(participants).toList
     getIdentifiersFromEnrolments(enrolments).intersect(participantIdentifiers).headOption
   }
+  @SuppressWarnings(Array("org.wartremover.warts.ImplicitParameter"))
```
[the remark this answers is not in the document]
Comment (jim-collins) [reply app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:141]: I was told to move them from the class to the methods.

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala [hunk 29]
```diff
@@ -127,6 +159,7 @@ class SecureMessageService @Inject()(repo: ConversationRepository, emailConnecto
   private def getIdentifiersFromParticipants(participants: Participants): NonEmptyList[Identifier] =
     participants.participants.map(p1 => p1.identifier)
 
+  @SuppressWarnings(Array("org.wartremover.warts.ImplicitParameter"))
```
[the remark this answers is not in the document]
Comment (jim-collins) [reply app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:162]: I was told to move them from the class to the methods.

[the remark this answers is not in the document]
Comment (OlegEfrem) [reply app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:162]: by whom :) ?

**Review by rwalpole — APPROVED** [review]

**Review by OlegEfrem — DISMISSED** [review]

**Review by elabeca — DISMISSED** [review]

File: app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala [hunk 30]
```diff
@@ -16,54 +16,89 @@
...
+    } yield res).value
+
+  private def validateCustomerParticipants(participants: GroupedParticipants)(errorCondition: => Boolean)(
+    implicit ec: ExecutionContext): EitherT[Future, SecureMessageError, Boolean] =
+    EitherT(Future {
+      if (errorCondition) {
+        Left(NoReceiverEmailError(s"Email lookup failed for: ${participants.customer.failure}"))
       } else {
-        Conflict("Duplicate of existing conversation")
+        Right(true)
```
[the remark this answers is not in the document]
Comment (jim-collins) [reply app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:66]: I will leave it for now I just want to see the back of this.

[the remark this answers is not in the document]
Comment (elabeca) [reply app/uk/gov/hmrc/securemessage/services/SecureMessageService.scala:66]: perfectly good with me, thanks

**Review by matthewgroom — APPROVED** [review]
